### PR TITLE
Firefox webextensions core

### PR DESCRIPTION
**Pull request by Hainish**
Core changes to WebExtension code to make it work with Firefox.  Reference the `done` column of https://github.com/EFForg/https-everywhere/projects/1

### CODE REVIEW

**Comment by Hainish** [comment]
@jsha: reverted 7dbab77ef08d86d8ca0219fa47d7fc54312d0ed9

**Comment by Hainish** [comment]
(As with #11792, just let me know when this looks good but don't merge.  I'll rebase to remove the original commit as well as the revert).

File: chromium/background.js [hunk 1]
```diff
@@ -618,3 +627,38 @@ chrome.runtime.onConnect.addListener(function (port) {
     });
   }
 });
+
+// This is necessary for communication with the popup in Firefox Private
+// Browsing Mode, see https://bugzilla.mozilla.org/show_bug.cgi?id=1329304
+chrome.runtime.onMessage.addListener(function(message, sender, sendResponse){
```
Comment (jsha) [inline chromium/background.js]: Can you confirm that this will only receive messages from background.js, not from content scripts or other extensions?

Comment (Hainish) [inline chromium/background.js]: That's right.  There is a separate IPC API for inter-extension, `chrome.runtime.onMessageExternal`.  See https://developer.chrome.com/extensions/messaging#external

File: chromium/popup.js [hunk 2]
```diff
@@ -65,11 +72,17 @@ function appendRuleLineToListDiv(ruleset, list_div) {
       break;
     }
   }
-  var xhr = new XMLHttpRequest();
-  try {
-    xhr.open("GET", favicon.src, true);
-    label.appendChild(favicon);
-  } catch (e) {}
+  if (favicon.src == "chrome://favicon/") {
+    favicon.src += "https://" + hostnames[ruleset.name];
```
Comment (koops76) [inline chromium/popup.js]: @Hainish @jsha WTF was this XMLHTTPRequest stuff?

File: chromium/popup.js [hunk 3]
```diff
@@ -65,11 +72,17 @@ function appendRuleLineToListDiv(ruleset, list_div) {
...
+  }
+
+  if (navigator.userAgent.match("Chrome")) {
+    var xhr = new XMLHttpRequest();
+    try {
+      xhr.open("GET", favicon.src, true);
```
Comment (koops76) [inline chromium/popup.js]: Remove this XMLHTTPRequest stuff. I'm not sure what it was supposed to do at all.

File: chromium/storage.js [hunk 4]
```diff
@@ -1,6 +1,10 @@
 var storage = chrome.storage.local;
```
Comment (koops76) [inline chromium/storage.js]: This entire file can be deleted and replaced with simple line in each file `storage` is used.

`const storage = chrome.storage.sync || chrome.storage.local`

Comment (Hainish) [inline chromium/storage.js]: Within for example the tor browser, `chrome.storage.sync` may exist but throw an error.  That's the reason for the `lastError` check.

Comment (koops76) [inline chromium/storage.js]: File a bug for Tor Browser then.

**Pull request merged**